File: 042.py
#   042
#   Trapping Rain Water
#   2015-05-18
#####################################################
class Solution:
    # @param {integer[]} height
    # @return {integer}
    def trap(self, height):
        n = len(height)
        if n == 0:
            return 0
        l = []
        r = []
        l.append(height[0])
        r.append(height[n-1])
        for i in range(1,n):
            j = n-1-i
            l.append(max(height[i], l[i-1]))
            r.insert(0, max(height[j], r[0]))
        sum = 0
        for i in range(1,n-1):
            roof = min(l[i], r[i])
            if height[i] < roof:
                sum += roof - height[i]
        return sum
        
#####################################################

sol = Solution()

height = [0,1,0,2,1,0,1,3,2,1,2,1]
print sol.trap(height)

File: 042.rb
# 042
# Trapping Rain Water
# 2015-05-18
#####################################################
# @param {Integer[]} height
# @return {Integer}
def trap(height)
  n = height.size
  return 0 if n == 0
  l = []
  r = []
  l[0] = height[0]
  r[n-1] = height[n-1]
  for i in (1..n-1)
    j = n-1-i
    l[i] = [height[i], l[i-1]].max
    r[j] = [height[j], r[j+1]].max
  end
  sum = 0
  for i in (1...n-1)
    roof = [l[i], r[i]].min
    sum += roof - height[i] if height[i] < roof
  end
  sum
end

#####################################################

height = [0,1,0,2,1,0,1,3,2,1,2,1]
puts trap(height)

